Making is_Mul LaTeX expressions consistent with Pretty Printer

arihantparsoya opened the pull request:
This is fix for issue #11275.
#### Output before this PR:

``` python
In [9]: var('x', positive=True)
Out[9]: x

In [10]: latex(exp(-x)*log(x))
Out[10]: '\\frac{1}{e^{x}} \\log{\\left (x \\right )}'

In [14]: var('x y')
Out[14]: (x, y)

In [15]: latex(1/(x + y)/2)
Out[15]: '\\frac{1}{2 x + 2 y}'

```
#### Output after this PR:

``` python
>>> var('x', positive=True)
x
>>> latex(exp(-x)*log(x))
'e^{- x} \\log{\\left (x \\right )}'
>>> var('x y')
(x, y)
>>> latex(1/(x + y)/2)
'\\frac{1}{2 \\left(x + y\\right)}'
```
- Tests are added.

arihantparsoya committed:
Improved fraction() function when exact is set to True
  This is fix for #11275. Previously, the function fraction() didn't distinguish between the term if exact is set or unset. Now, when exact is set, the function prevents terms with negative exponents to be pushed to the denominator when their exact value is unknown .
 In line 982, the expression is preventing simplification of the expression when exact is set to true, such as : 2*(x+y) should remain 2*(x+y) but not (2*x + 2*y). This type of evaluation is crutial in printing the expression better and helps LaTeX printer stay in consistent with pretty printer.

parsoyaarihant commented:
@asmeurer , can you review this?

smichr commented on sympy/simplify/radsimp.py:
```diff
@@ -950,15 +950,22 @@ def fraction(expr, exact=False):
         if term.is_commutative and (term.is_Pow or term.func is exp):
             b, ex = term.as_base_exp()
-            if ex.is_negative:
+            if not exact and ex.is_negative:
```
I recommend that you merge these two if-blocks:

```
if ex.is_negative:
    if ex is S.NegativeOne: ...
    elif exact: ...
        if ex.is_constant:...
        else:
    else: ...
```

parsoyaarihant replied:
@smichr , thanks.

asmeurer commented:
Since you are changing `fraction`, can you add some tests for `fraction` itself as well.

asmeurer commented:
@smichr any objections to the `fraction` change?

smichr commented:
> objections...

I will look more closely tomorrow afternoon.

parsoyaarihant commented:
@smichr , can you review this?

smichr commented on sympy/printing/tests/test_latex.py:
```diff
@@ -56,6 +56,8 @@ def test_latex_basic():
 
     assert latex(2*x*y) == "2 x y"
     assert latex(2*x*y, mul_symbol='dot') == r"2 \cdot x \cdot y"
+    assert latex(exp(-x)*log(x)) == r"e^{- x} \log{\left (x \right )}"
```
<neutral question>: This is not different than master; what is the purpose?

parsoyaarihant replied:
This test was written to address the original issue. Since it now relates to tests for `fraction()`. I will remove them.

smichr commented on sympy/simplify/tests/test_radsimp.py:
```diff
@@ -393,6 +394,8 @@ def test_fraction():
     assert fraction(exp(-n)) == (exp(-n), 1)
 
+    p = symbols('p', positive=True)
+    assert fraction(exp(-p)*log(p), exact=True) == (exp(-p)*log(p), 1)
```
If this change is made then the docstring should be updated. It says, 

```
If we know nothing about sign of some exponent and 'exact'
flag is unset, then structure this exponent's structure will
be analyzed and pretty fraction will be returned:
```

But in this case we *do* know something about the sign of the exponent. My read of the current docstring is that `exact` only applies to vanilla exponents.

parsoyaarihant replied:
* I referred to the docstring while making this PR. Since the docstring only refers to the case when `exact` is `unset`. It doesnt talk about if `exact=True`.

* The way I improved `exact=True` case is that it now applies to unknown exponents whose values is not definite. 

* I think we should add docstring for `exact=True` case keeping the previous docstrings same.

smichr replied:
> I referred to the docstring while making this PR. Since the docstring only refers to the case when exact is unset. It doesnt talk about if exact=True.

OK we have four cases:

1. exact and vanilla (sign unknown) exponent -- your case
2. not exact and vanilla -- docstring case
3. exact and not vanilla - not covered by docstring; should `exact` take precedence?
4. not exact and not vanilla - pretty result

smichr commented:
> can you review this

Busy overtime yesterday and driving til midnight. Sorry for the delay. I only have the two comments, one of which is a reservation to current changes.

smichr commented:
Case 3 does not need to be addressed by this PR. So I am +1 as is.

smichr commented:
@asmeurer , if you're ok with this, go ahead and commit.

parsoyaarihant commented:
@smichr @asmeurer , ping.

asmeurer commented:
Can you write a better commit message?

asmeurer commented:
Your commit message only mentions the fraction change, not the printer change. I would suggest breaking the two changes into two commits. 

Also, more than one line, explaining why the change was made (for instance, taking the justifications from the discussion here) would help a lot. The idea behind commit messages is if someone comes across the change later and wants to know why the change was made they can see it from the commit message.

parsoyaarihant commented:
I made the necessary changes. Please give your feedback on the commit message.

I will make sure to write complete commit message from now.

asmeurer commented:
That looks good. Were you planning on readding another commit for the latex printer change?

parsoyaarihant commented:
@asmeurer ,  previously I added some tests for latex printer but after discussion with @smichr , those tests were removed since they are direct consequence of function `fraction()` (Discussion related to those tests cannot be seen now since the all commits are squashed and the test file is removed).

I referenced latex printer in my last commit discription hence I felt that another commit won't be necessary.

asmeurer commented:
I would like to see tests for the printer directly, since that will test that the format is the same as the pretty printer.

arihantparsoya committed:
Added tests for LaTeX

Added tests for latex-printer directly.

This is done to make sure that LaTeX printing format is same as pretty
printer.

parsoyaarihant commented:
@asmeurer , ping.

asmeurer commented:
Don't remove the fraction tests. We should have both.

parsoyaarihant commented:
@asmeurer , tests for `fraction()` are added.

parsoyaarihant commented:
@asmeurer , ping.
